File: sections/28_dp_i_mixed_contest/problems/g_hamiltonian_route/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<vector<ll>> weight(n, vector<ll>(n));
    for (auto& row : weight)
        for (ll& value : row)
            cin >> value;

    const ll INF = 1LL << 62;
    const int mask_count = 1 << n;
    vector<vector<ll>> dp(mask_count, vector<ll>(n, INF));
    dp[1][0] = 0;

    for (int mask = 1; mask < mask_count; ++mask) {
        for (int u = 0; u < n; ++u) {
            if (dp[mask][u] == INF)
                continue;

            for (int v = 0; v < n; ++v) {
                if ((mask >> v & 1) || weight[u][v] < 0)
                    continue;

                int next_mask = mask | (1 << v);
                dp[next_mask][v] = min(
                    dp[next_mask][v],
                    dp[mask][u] + weight[u][v]
                );
            }
        }
    }

    ll answer = dp[mask_count - 1][n - 1];
    cout << (answer == INF ? -1 : answer) << '\n';
}

File: sections/28_dp_i_mixed_contest/problems/g_hamiltonian_route/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    values = data[1:]
    weight = [
        values[row * n:(row + 1) * n]
        for row in range(n)
    ]

    infinity = 10**30
    mask_count = 1 << n
    dp = [[infinity] * n for _ in range(mask_count)]
    dp[1][0] = 0

    for mask in range(mask_count):
        for u in range(n):
            if dp[mask][u] == infinity:
                continue

            for v in range(n):
                if mask >> v & 1 or weight[u][v] < 0:
                    continue

                next_mask = mask | (1 << v)
                dp[next_mask][v] = min(
                    dp[next_mask][v],
                    dp[mask][u] + weight[u][v],
                )

    answer = dp[mask_count - 1][n - 1]
    print(-1 if answer == infinity else answer)


if __name__ == "__main__":
    main()
